File: code/library/src/main/java/org/fbme/lib/iec61499/stringify/SystemPrinter.java
package org.fbme.lib.iec61499.stringify;


import org.fbme.lib.iec61499.declarations.*;
import org.fbme.lib.iec61499.fbnetwork.FBNetwork;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;

public final class SystemPrinter extends DeclarationPrinterBase<SystemDeclaration> {

    public SystemPrinter(SystemDeclaration declaration) {
        super(declaration, "System");
    }

    @Override
    protected void printDeclarationBody(@NotNull Element element) {
        printApplications(element);
        printDevices(element);
        printMappings(element);
        printSegments(element);
        printLinks(element);
    }

    private void printApplications(Element element) {
        for (ApplicationDeclaration applicationDeclaration : myElement.getApplications()) {
            element.addContent(new ApplicationPrinter(applicationDeclaration).print());
        }
    }

    private static class ApplicationPrinter extends DeclarationPrinterBase<ApplicationDeclaration> {

        public ApplicationPrinter(ApplicationDeclaration applicationDeclaration) {
            super(applicationDeclaration, "Application");
        }

        @Override
        protected void printDeclarationBody(@NotNull Element element) {
            element.addContent(new SubappNetworkPrinter(myElement.getNetwork()).print());
        }
    }

    private void printDevices(Element element) {
        for (DeviceDeclaration deviceDeclaration : myElement.getDevices()) {
            element.addContent(new DevicePrinter(deviceDeclaration).print());
        }
    }

    private static class DevicePrinter extends DeclarationPrinterBase<DeviceDeclaration> {

        public DevicePrinter(DeviceDeclaration deviceDeclaration) {
            super(deviceDeclaration, "Device");
        }

        @Override
        protected void printDeclarationBody(@NotNull Element element) {
            element.setAttribute("Type", myElement.getTypeReference().getPresentation());
            ParameterAssignmentPrinter.printAll(myElement.getParameters(), element);
            for (ResourceDeclaration resourceDeclaration: myElement.getResources()) {
                element.addContent(new ResourcePrinter(resourceDeclaration).print());
            }
            FBNetwork network = myElement.getNetwork();
            if (network != null) {
                element.addContent(new FBNetworkPrinter<>(network).print());
            }
        }
    }

    private void printMappings(Element element) {
        for (Mapping mapping: myElement.getMappings()) {
            Element mappingElement = new Element("Mapping");
            mappingElement.setAttribute("From", mapping.getApplicationFBReference().getPresentation());
            mappingElement.setAttribute("To", mapping.getResourceFBReference().getPresentation());
            element.addContent(mappingElement);
        }
    }


    private void printSegments(Element element) {
        for (SegmentDeclaration segmentDeclaration: myElement.getSegments()) {
            element.addContent(new SegmentPrinter(segmentDeclaration).print());
        }
    }

    private static class SegmentPrinter extends DeclarationPrinterBase<SegmentDeclaration> {

        public SegmentPrinter(SegmentDeclaration segmentDeclaration) {
            super(segmentDeclaration, "Segment");
        }

        @Override
        protected void printDeclarationBody(@NotNull Element element) {
            element.setAttribute("Type", myElement.getTypeReference().getPresentation());
            ParameterAssignmentPrinter.printAll(myElement.getParameters(), element);
        }
    }

    private void printLinks(Element element) {
        for (Link link: myElement.getLinks()) {
            Element linkElement = new Element("Link");
            linkElement.setAttribute("CommResource", link.getResourceReference().getPresentation());
            linkElement.setAttribute("SegmentName", link.getSegmentReference().getPresentation());
            element.addContent(linkElement);
        }
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/stringify/SystemPrinter.kt
package org.fbme.lib.iec61499.stringify

import org.fbme.lib.iec61499.declarations.ApplicationDeclaration
import org.fbme.lib.iec61499.declarations.DeviceDeclaration
import org.fbme.lib.iec61499.declarations.SegmentDeclaration
import org.fbme.lib.iec61499.declarations.SystemDeclaration
import org.jdom.Element

class SystemPrinter(declaration: SystemDeclaration) :
    DeclarationPrinterBase<SystemDeclaration>(declaration, "System") {
    override fun printDeclarationBody(element: Element) {
        printApplications(element)
        printDevices(element)
        printMappings(element)
        printSegments(element)
        printLinks(element)
    }

    private fun printApplications(element: Element) {
        for (applicationDeclaration in this.element.applications) {
            element.addContent(ApplicationPrinter(applicationDeclaration).print())
        }
    }

    private class ApplicationPrinter(applicationDeclaration: ApplicationDeclaration) :
        DeclarationPrinterBase<ApplicationDeclaration>(applicationDeclaration, "Application") {
        override fun printDeclarationBody(element: Element) {
            element.addContent(SubappNetworkPrinter(this.element.network).print())
        }
    }

    private fun printDevices(element: Element) {
        for (deviceDeclaration in this.element.devices) {
            element.addContent(DevicePrinter(deviceDeclaration).print())
        }
    }

    private class DevicePrinter(deviceDeclaration: DeviceDeclaration) :
        DeclarationPrinterBase<DeviceDeclaration>(deviceDeclaration, "Device") {
        override fun printDeclarationBody(element: Element) {
            element.setAttribute("Type", this.element.typeReference.presentation)
            ParameterAssignmentPrinter.printAll(this.element.parameters, element)
            for (resourceDeclaration in this.element.resources) {
                element.addContent(ResourcePrinter(resourceDeclaration).print())
            }
            val network = this.element.network
            if (network != null) {
                element.addContent(FBNetworkPrinter(network).print())
            }
        }
    }

    private fun printMappings(element: Element) {
        for (mapping in this.element.mappings) {
            val mappingElement = Element("Mapping")
            mappingElement.setAttribute("From", mapping.applicationFBReference.presentation)
            mappingElement.setAttribute("To", mapping.resourceFBReference.presentation)
            element.addContent(mappingElement)
        }
    }

    private fun printSegments(element: Element) {
        for (segmentDeclaration in this.element.segments) {
            element.addContent(SegmentPrinter(segmentDeclaration).print())
        }
    }

    private class SegmentPrinter(segmentDeclaration: SegmentDeclaration) :
        DeclarationPrinterBase<SegmentDeclaration>(segmentDeclaration, "Segment") {
        override fun printDeclarationBody(element: Element) {
            element.setAttribute("Type", this.element.typeReference.presentation)
            ParameterAssignmentPrinter.printAll(this.element.parameters, element)
        }
    }

    private fun printLinks(element: Element) {
        for (link in this.element.links) {
            val linkElement = Element("Link")
            linkElement.setAttribute("CommResource", link.resourceReference.presentation)
            linkElement.setAttribute("SegmentName", link.segmentReference.presentation)
            element.addContent(linkElement)
        }
    }
}
